fix: NetworkAnimator does not update animator's parameters unless the animator's layer changes [MTT-3492]

NoelStephensUnity opened the pull request:
- Animator parameter values are only synchronized if the animator layer changes or a transition takes place.
  - Fix: Parameter values should be synchronized if/when they change
- NetworkAnimator does not synchronize late-joining players to current animation states.
  - Fix: Late-joining clients should be fully synchronized with the current NetworkAnimator state

This PR fixes these issues:
[MTT-3492](https://jira.unity3d.com/browse/MTT-3492)
[MTT-2644](https://jira.unity3d.com/browse/MTT-2644)

## Changelog
- Fixed: Animator parameter values are only synchronized if the animator layer changes or a transition takes place.
- Fixed: NetworkAnimator does not synchronize late-joining players to current animation states.

## Testing and Documentation
- Includes 3 integration tests for both server and client owners
  - Verifies parameters are updated without changing layers or transitioning 
  - Verifies triggers are properly updated 
  - Verifies late joining client is synchronized to current animation states
- Includes fixes and updates to the manual NetworkAnimator test and related test project assets

## About Files Changed:
- Only 1 of the 32 files changed are within the components assembly (NetworkAnimator).
  - The remaining 31 files changed are all pertaining to manual and integration testing.

## Additional Information:
NetworkAnimator should now only write what has changed and not everything about the animation state --unless-- it is synchronizing late joining clients.

- Triggers should still be set via the NetworkAnimator.SetTrigger to get "best visual result".
  - Setting the trigger directly on the Animator still works, but loses a portion of the initial transition blending from the source state to the destination state (so it will look like it "pops" to the exit time of the transition).
  - When synchronizing late joining players and the Animator is in the middle of a transition, the late joining player will be synchronized to the next (destination) state.
- States are now only updated if their normalized time is < 1.0f (i.e. if a state's normalized time is >= 1.0f then it is no longer playing or about to loop back to 0.0f).
  - Note: The normalized time threshold takes speed and the speed multiplier into consideration

NoelStephensUnity committed:
fix

Fix for animators that have one layer and handle most animation state changes via properties.

NoelStephensUnity committed:
test and style

Adding an integration test to validate the changes.
Updated comments in NetworkAnimator.

NoelStephensUnity committed:
style

MTT-3492
Remove namespace references I was not using.

NoelStephensUnity committed:
style

only applying changes required for asmdef file.

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -138,15 +156,29 @@ public override void OnNetworkSpawn()
                         case AnimatorControllerParameterType.Float:
                             var value = m_Animator.GetFloat(cacheParam.Hash);
                             UnsafeUtility.WriteArrayElement(cacheParam.Value, 0, value);
+                            // Track this parameter
+                            if (!m_ParameterStates.ContainsKey(i))
+                            {
+                                m_ParameterStates.Add(i, new ParameterState() { CurrentStateFloat = value, AnimatorControllerParameterType = AnimatorControllerParameterType.Float });
+                            }
                             break;
```
I'm looking at 2 different approaches in one place here.
what's the problem you're trying to solve here that can't be achieved with existing `UnsafeUtility.WriteArrayElement` approach?
I just don't like seeing Dictionaries everywhere and being thrown at every problem we have.
I'm just curious whether this issue could be solved more simple or not.

NoelStephensUnity replied:
Basically the NetworkAnimator will only update an Animator's parameters when an animation is going through a transition and/or the animator's layer has changed.  While this works for when you are transitioning or have changed the layer, it misses all other cases where you are just changing the Animator's parameter values (i.e. no longer transitioning and not changing the animator's layer).
This table was a quick way to detect if a parameter has changed and if so it does a full update.

Really, I should bring this back to draft and make some optimization adjustments where it will only write what has changed and not everything each time it serializes.

But to answer the question of what that is for:  It tracks the current parameter values to be able to determine if one has changed.

NoelStephensUnity replied:
See my comments below, I now see what you were talking about and completely removed the dictionary.

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
update and test

Removed the duplicate tracking for parameters.
Added the ability to make it owner or server authoritative.
Included triggers in this update.
Added test for triggers.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity commented:
@ThusWroteNomad I made a few additional modifications that removes the additional list and provides the ability to be owner or server authoritative.
We still could greatly optimize the way we send updates, but I think we should create a separate PR for the optimizations.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -239,6 +266,55 @@ private bool CheckAnimStateChanged(out int stateHash, out float normalizedTime,
                 }
             }
 
+            // Whether the layer has changed or not, we always check to see if parameters have changed and send updates if they have.
```
I think for a future PR we could create a full index of what changed and then send only what has changed.
For this PR I kept it relatively simple since updating meant it would send all values anyway.

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
update

removing unused parameter state class

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
update and fix

This updates the manual animation test so we can visually see NetworkAnimator working properly.
This fixes the issue where late joining clients would not be updated to the NetworkAnimator's state.
This fixes some test project asset related issues including to make sure we are always set to run in the background and have a target fps of 120.

NoelStephensUnity committed:
update and fix

This fixes the issue where it would sometimes not clean up/dispose the "cache" native array.
This improves the client synchronization (as best for the current way that we write information).

Manual Test Updates:
This fixes an issue with scale as well as getting other players to join and distribute evenly.
This fixes the issue with the prefab asset itself (scale issue impacted pulse).
Increased send/receive buffer sizes (we send a lot of data for animations currently)

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -89,19 +93,28 @@ static AnimationParamEnumWrapper()
         }
 
-        public override void OnDestroy()
+        protected void CleanUp()
```
can we just "not" expose this API to keep this PR backportable to `release/1.0.0`?

NoelStephensUnity committed:
update

Making CleanUp private.

NoelStephensUnity committed:
fix

This lost its parameters for some reason...

NoelStephensUnity committed:
fix

Missed adding the trigger state back.

NoelStephensUnity committed:
test

This verifies that late joining clients get synchronized to the current NetworkAnimator state.
This also fixes the issue where stopping a client during an integration test always meant that you wanted to delete it.  Now, you have the option to not destroy in the event you want to stop a client, do some stuff, and then reconnect the client.

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
test and style

Minor tweak and additional check that the AnimatorTestHelper instances equals the TotalClient count (more of a sanity check).
Added some comments and removed some LFs.

NoelStephensUnity committed:
Test and Style

I hooked up the TriggerTest to the wrong state.  Fixed issue by creating a new state that is activated by the test trigger.
Changing the warning to an actual error in TriggerTest (i.e. if that happens the test should fail).

NoelStephensUnity committed:
Update

This includes several adjustments to improve performance.
-Parameters are now sent separately
-New internal NetworkAnimatorStateChangeHandler that changes when:
  -It detects changes to the Animator
  -The server-side client forwarded RPC messages are processed here
-Removed the OwnerAuthoritative public bool

This also include integration test adjustments in order to be able to create and join a new client on the fly during a test.

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
fix

This got test running locally in a stand alone testrunner build.
Going to give this a shot in Yamato land.

NoelStephensUnity committed:
wip

Sorting out the issue with server-side owner late joining client synchronization.  This version works with all tests ==but== the server-side owner late joining client portion... still trying to find this bug.

NoelStephensUnity committed:
fix

This reverts Triggers back to being set by an RPC (better to separate for now).
This also fixes minor issues with a player joining right in the middle of a transition.
This also fixes issue where trigger paths were not being set when transitioning (still could use more work).
This fixes the issue with the NetworkAnimator failing on the late joining player test where the late joining player could join while an Animator is in the middle of a transition.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
style

NoelStephensUnity committed:
update

This improves upon the synchronization of a late joining client when connected while an animation is in the middle of a transition (it doesn't fix the issue completely, but is closer to being synchronized with the source animation)

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
update

Removing VerboseDebug information

NoelStephensUnity committed:
Test

Added a small profiling test to profile where we spend time.
(sending RPCs)

NoelStephensUnity committed:
update - optimize

Optimizing how we send parameter updates.  Now we only send the parameters that changed.
This also avoids some smaller allocations by reusing a ClientRpcParams instance and a List<ulong> to hold the client ids to send to.

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

NoelStephensUnity committed:
style

updating a missed whitespace

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -80,40 +236,89 @@ private struct AnimationParamEnumWrapper
         }
 
-        public override void OnDestroy()
+        private void CleanUp()
```
nit
```suggestion
        private void Cleanup()
```

NoelStephensUnity replied:
Applied suggestion

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -142,38 +347,129 @@ public override void OnNetworkSpawn()
+            }
+        }
+
+        private void Server_OnClientConnectedCallback(ulong playerId)
```
```suggestion
        private void ServerOnClientConnectedCallback(ulong playerId)
```

0xFA11 replied:
we shouldn't use underscore character here, maybe rename the method as something else if `ServerOnClient...` doesn't look good to you?

NoelStephensUnity replied:
I can just rename it to OnClientConnectedCallback...

NoelStephensUnity replied:
Applied alternate to suggestion, but accomplishes the same goal.

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -187,29 +483,120 @@ private void FixedUpdate()
+        /// <summary>
+        /// Checks if any of the Animator's states have changed
+        /// </summary>
+        unsafe private bool CheckAnimStateChanged(out int stateHash, out float normalizedTime, int layer)
```
nit
```suggestion
        private unsafe bool CheckAnimStateChanged(out int stateHash, out float normalizedTime, int layer)
```

NoelStephensUnity replied:
Applied suggestion

0xFA11 reviewed (approved):
some nits but overall, looks good to me

0xFA11 reviewed (changes requested):
oh actually, we need a changelog entry nevertheless :)

NoelStephensUnity committed:
update

MTT-3492 MTT-2644
Updating what was fixed in the change log.

NoelStephensUnity committed:
update

Minor tweaks suggested by Fatih.

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-single-layer

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -25,6 +25,12 @@ Additional documentation and release notes are available at [Multiplayer Documen
 - Fixed issues when multiple `ConnectionApprovalCallback`s were registered (#1972)
 - `FixedString` types can now be used in NetworkVariables and RPCs again without requiring a `ForceNetworkSerializeByMemcpy<>` wrapper (#1961)
 - Fixed endless dialog boxes when adding a `NetworkBehaviour` to a `NetworkManager` or vice-versa. (#1947)
+- Fixed `NetworkAnimator` issue where it was only synchronizing parameters if the layer or state changed or was transitioning between states. (#1946)
+- Fixed `NetworkAnimator` issue where when it did detect a parameter had changed it would send all parameters as opposed to only the parameters that changed. (#1946)
+- Fixed `NetworkAnimator` issue where it was not always disposing the `NativeArray` that is allocated when spawned. (#1946)
+- Fixed `NetworkAnimator` issue where it was not taking the animation speed or state speed multiplier into consideration. (#1946)
+- Fixed `NetworkAnimator` issue where it was not properly synchronizing late joining clients if they joined while `Animator` was transitioning between states. (#1946)
+- Fixed `NetworkAnimator` issue where the server was not relaying changes to non-owner clients when a client was the owner. (#1946)
```
super cool!
